src/kernel/parsetree_cache.ml

```diff
@@ -0,0 +1,82 @@
+let { Logger.log } = Logger.for_section "Parsetree_cache"
+
+module Mreader_cache = struct
+  type t = { result : Mreader.result; was_cached : bool }
+  type cache = { last_digest : Msource.Digest.t; last_result : Mreader.result }
+
+  let cache = ref None
+
+  let with_cache (source, pp_result) f =
+    let title = "Mreader cache" in
+    let result, was_cached =
+      match (!cache, pp_result) with
+      | Some { last_digest; last_result }, None
+        when Msource.Digest.(equal last_digest @@ make source) ->
+          log ~title "mreader result was re-used";
+          (last_result, true)
+      | _ ->
+          log ~title "source was parsed";
+          (f (source, pp_result), false)
+    in
+    cache :=
+      Some { last_digest = Msource.Digest.make source; last_result = result };
+    { result; was_cached }
+end
+
+module Ppx_cache = struct
+  type t = Mreader.parsetree
+
+  module Fingerprint = struct
+    type t = { binary_id : File_id.t; args : string list; workdir : string }
+
+    let make ~binary ~args ~workdir =
+      let qualified_binary = Misc.add_segment ~dir:workdir ~seg:binary in
+      Result.map (fun binary_id -> { binary_id; args; workdir })
+      @@ File_id.get_res qualified_binary
+
+    let equal { binary_id = b1; args = a1; workdir = w1 }
+        { binary_id = b2; args = a2; workdir = w2 } =
+      File_id.check b1 b2 && List.equal String.equal a1 a2 && String.equal w1 w2
+  end
+
+  type cache = { fingerprints : Fingerprint.t list; result : Mreader.parsetree }
+
+  let cache = ref None
+
+  let with_cache ~source_was_cached (config : Mconfig.t) f =
```

Comment:
No, I haven't thaught about it very much tbh. But I do think it's a good idea. It would make it more coherent. I don't have time for it today anymore. If you think it's important, I'll try to find time. I don't think it would be that difficult.

Reply:
I've functorized the two caches now. It's a bit longer than before, but indeed more coherent and easier to read IMO. Thanks for the idea!

Btw, with this new abstraction, this all-or-nothing cache mechanism can be used for any phase of the pipeline. However, the typer phase already has a way more elaborate cache and the computation phase in the majority of cases isn't a bottleneck. Still, it might be interesting to look into using it for concrete cases of the computation phase at some point, such as `complete-prefix` queries (I haven't thought this through, though; it's just an idea).